### PR TITLE
debian openvpn has dedicated user and group, so rather use those than…

### PR DESCRIPTION
**Pull request by corbolais**
… nobody:nogroup (as do _many_ other packages and thus all having access to such owned files).

Signed-off-by: corbolais <corbolais@gmail.com>

### CODE REVIEW

**Comment by orazioedoardo** [comment]
I don't have the openvpn group on Debian:
```
$ dpkg -s openvpn | grep '^Status:'
Status: install ok installed
$ grep openvpn /etc/group
$ 
```

**Comment by corbolais** [comment]
> I don't have the openvpn user nor group on Debian:

that's interesting. would you mind posting `cat /etc/os-release`?

I've got a couple of opvn servers here, all of them feature the user/group combo.

**Comment by orazioedoardo** [comment]
```
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**Comment by corbolais** [comment]
Curious
Seems to be a bullseye/sid thing, too.
All openvpn servers here are bullseye/sid:

```
PRETTY_NAME="Debian GNU/Linux bullseye/sid"
NAME="Debian GNU/Linux"
ID=debian
```

```
getent passwd openvpn
openvpn:x:123:130::/var/lib/openvpn:/usr/sbin/nologin
```

**Comment by orazioedoardo** [comment]
Same on Raspbian by the way, can you check Ubuntu too?

**Comment by corbolais** [comment]
```
root:~# grep openvpn /etc/passwd
nm-openvpn:x:124:136:NetworkManager OpenVPN,,,:/var/lib/openvpn/chroot:/usr/sbin/nologin
root:~# cat /etc/os-release 
NAME="Ubuntu"
VERSION="19.04 (Disco Dingo)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 19.04"
VERSION_ID="19.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=disco
UBUNTU_CODENAME=disco
```

Do you want me to add this user to the PLAT=Ubuntu path?

**Comment by orazioedoardo** [comment]
> Do you want me to add this user to the PLAT=Ubuntu path?

Isn't that user related just to the Network Manager integration with openvpn? I would just create an 'openvpn' user for every platform, so you don't need [this test](https://github.com/pivpn/pivpn/pull/877/files#diff-299d9b30c952310aace4aba77c698874R1295) (by the way no nm-openvpn nor openvpn user on ubuntu 18.04).

**Comment by orazioedoardo** [comment]
Or maybe check if ‘nm-openvpn’ exists. If it exists, use it, otherwhise create ‘openvpn’. But what would happen if the user uninstalls or replaces Network Manager? It will remove the ‘nm-openvpn’ user thus breaking the openvpn server.

**Comment by corbolais** [comment]
> so you don't need [this test](https://github.com/pivpn/pivpn/pull/877/files#diff-299d9b30c952310aace4aba77c698874R1295) (by the way no nm-openvpn nor openvpn user on ubuntu 18.04).

That would be my favourite, yes.

**Comment by orazioedoardo** [comment]
Can you add some code to remove the openvpn user if the user chooses to uninstall the openvpn package (in the uninstall.sh script)?

**Comment by orazioedoardo** [comment]
Regarding [this](https://github.com/pivpn/pivpn/pull/877/commits/729674595bb9071fcbf06d4384c72fc456b4dcd4#diff-409efba6db37b831b75997c11a4c9ec4R111), maybe use different naming when creating those files at the time of installation? What's the problem [here](https://github.com/pivpn/pivpn/pull/877/commits/729674595bb9071fcbf06d4384c72fc456b4dcd4#diff-409efba6db37b831b75997c11a4c9ec4R138)?

**Comment by 4s3ti** [comment]
Sorry i haven't read the whole topic ... quite tight on time ... and almost finishing with the website in the new domain .. however ... not entirely sure if its related or not wanted to give an heads up. 

Openvpn Drops permissions to nobody:Nogroup due to security reasons .. please be careful to don't go against it!

**Comment by orazioedoardo** [comment]
@4s3ti Actually he's right. Dropping permissions for security reasons is ok but it's effectiveness is reduced if other processes drop to the same users. Makes sense to drop permissions to a specific user only used by openvpn.

**Comment by 4s3ti** [comment]
@orazioedoardo aight!! i kinda just saw the notification and wanted to give a heads up first! =) 
it has been tight on time lately but soon will get back on code and be able to allocate regular time to it! =)

**Comment by corbolais** [comment]
> Regarding [this](https://github.com/pivpn/pivpn/pull/877/commits/729674595bb9071fcbf06d4384c72fc456b4dcd4#diff-409efba6db37b831b75997c11a4c9ec4R111), maybe use different naming when creating those files at the time of installation? 

Yes, that's the way to go. Didn't modify on short notice but didn't want to let it pass uncommented either. So, solution to this would be a proper naming scheme. Which IMO falls right in line with a refactor of commonly used functions, magic numbers, general double quoting when using i.e. whiptail and coding style (line spacing and indentation levels/chars in particular). Then again, the naming scheme must be consistent for install and uninstall routines, again, requiring some persistant config storage.

> What's the problem [here](https://github.com/pivpn/pivpn/pull/877/commits/729674595bb9071fcbf06d4384c72fc456b4dcd4#diff-409efba6db37b831b75997c11a4c9ec4R138)?

Basically the same issue: locally user-modified files unconditionally removed. Either use dedicated pipvpn config files where feasible or really only remove stuff pivpn added.

Same goes for iptables/ufw rules. I already improved on this in install.sh regarding in-place modification and forced *nat table addition. Generally, it is recommended to just avoid in-place modification using sed, it is way too fragile and you don't want to mess up users' configs, even less with fw configs in particular.

**Comment by orazioedoardo** [comment]
> Basically the same issue: locally user-modified files unconditionally removed. Either use dedicated pipvpn config files where feasible or really only remove stuff pivpn added.

Agree. It's a bit mitigated by the fact that those files are removed only if pivpn itself installed unattended-upgrades (see [here](https://github.com/pivpn/pivpn/blob/4466f1503c9797808d49d8cb55063bec5a79deed/auto_install/install.sh#L27-L30)).

This could be further expanded by copying pivpn specific setting from files in the git repo instead of inlining them into the script, what do you think?

**Comment by corbolais** [comment]
For what it's worth, maybe it should be considered globally implementing the use of `install(1)`.

**Comment by corbolais** [comment]
> Agree. It's a bit mitigated by the fact that those files are removed only if pivpn itself installed unattended-upgrades (see [here](https://github.com/pivpn/pivpn/blob/4466f1503c9797808d49d8cb55063bec5a79deed/auto_install/install.sh#L27-L30)).

I'm having a hard time seeing this as an effective mitigation with respect to the underlying issue.

> 
> This could be further expanded by copying pivpn specific setting from files in the git repo instead of inlining them into the script, what do you think?

Absolutely.

**Comment by orazioedoardo** [comment]
> I'm having a hard time seeing this as an effective mitigation with respect to the underlying issue.

If the issue is "don't remove files that the user may have already modified for its own purposes" then it's a mitigation, since if pivpn installed the package those files where not there, so it's safe to remove them.

> For what it's worth, maybe it should be considered globally implementing the use of `install(1)`.

If you are willing to implement this and subsequent changes to improve the script, let me know and I will keep these pull requests open until you are confident they can be merged to test.

**Comment by orazioedoardo** [comment]
> If the issue is "don't remove files that the user may have already modified for its own purposes" then it's a mitigation, since if pivpn installed the package those files where not there, so it's safe to remove them.

Ok well there is a case where the user may modify pivpn installed configs after the package is installed.

**Comment by corbolais** [comment]
> If the issue is "don't remove files that the user may have already modified for its own purposes" then it's a mitigation, since if pivpn installed the package those files where not there, so it's safe to remove them.

OK, if this is true for every single file in question.

> > For what it's worth, maybe it should be considered globally implementing the use of `install(1)`.
> 
> If you are willing to implement this and subsequent changes to improve the script, let me know and I will keep these pull requests open until you are confident they can be merged to test.

I can do so. Meanwhile, there's no need to keep this PRs open. They can be merged into test as-is.

Any future improvements go in independently. It will be a rather lengthy process to completely overhaul the scripts, so it will happen over a stretch of time.

**Comment by corbolais** [comment]
> > If the issue is "don't remove files that the user may have already modified for its own purposes" then it's a mitigation, since if pivpn installed the package those files where not there, so it's safe to remove them.
> 
> Ok well there is a case where the user may modify pivpn installed configs after the package is installed.

ACK

**Comment by orazioedoardo** [comment]
> I can do so. Meanwhile, there's no need to keep this PRs open. They can be merged into test as-is.

I will merge https://github.com/pivpn/pivpn/pull/876, https://github.com/pivpn/pivpn/pull/880 and https://github.com/pivpn/pivpn/pull/878. Not https://github.com/pivpn/pivpn/pull/879 as it overlaps with https://github.com/pivpn/pivpn/pull/880.

**Comment by corbolais** [comment]
> > I can do so. Meanwhile, there's no need to keep this PRs open. They can be merged into test as-is.
> 
> I will merge #876, #880 and #878. Not #879 as it overlaps with #880.
Not sure what happened w/ 47c84e6a45a9142800f4669cf4a5e72ac0c36122, it ended up identically in two PRs, of which only #879 was intended.

**Comment by orazioedoardo** [comment]
#880 has all the changes for the uninstall.sh script from #879 plus changes for the install.sh. Are you discarding the latter?

**Comment by corbolais** [comment]
Uh. I will investigate this. Sorry. Lots of PRs on top of each other or depending on one or another.

**Comment by corbolais** [comment]
> #880 has all the changes for the uninstall.sh script from #879 plus changes for the install.sh. Are you discarding the install.sh? (I will fix the merge conflict If you choose #880)

Ok, the way I meant this is: I originally intended commit 47c84e6a45a9142800f4669cf4a5e72ac0c36122 for PR #879, with #879 being the PR for `update.sh`. 

However, additionally, commit 47c84e6a45a9142800f4669cf4a5e72ac0c36122 somehow erroneously ended up in PR #880. 

So, as it is now, it seems best to drop PR #879 and just use PR #880 as commit 47c84e6a45a9142800f4669cf4a5e72ac0c36122 is already part of #880.

Does that make sense to anyone else? ;-)

Changes have been merged.

Thanks.
